### ISSUE TITLE
Format tags count towards truncating text

### ISSUE THREAD

**Issue by nunocastromartins** [jaagr/polybar#531]
Consider the following module:
```
[module/test]
type = custom/script
exec = echo "%{B#660000}this text has 27 characters"
label-maxlen = 27
```

It shows the following:
![2017-04-12-191719_111x29_scrot](https://cloud.githubusercontent.com/assets/8314616/24972905/e3d087e8-1fb4-11e7-95cb-44f5ea88de21.png)

I believe the format tags shouldn't be considered when truncating the text.

**Comment by unode** [comment]
Possibly related to this, formatting tags that define actions are disabled on truncated content:

```
[module/mpd]
type = internal/mpd

label-song = %{A3:sonata:}%artist% - %title%%{A}
label-song-maxlen = 30
label-song-ellipsis = true
```

Is not clickable and sonata is never launched. However if:

```
label-song-maxlen = 100
```

everything works as expected, unless the song title is really long.

I propose relabeling as a `bug`.

**Comment by Lomadriel** [comment]
Yeah indeed, if the string is truncated then the `%{A}` that close the clickable area is mising and thus ignored the whole action is ignored.
I will try to do something.

**Comment by patrick96** [comment]
@Lomadriel what did you have in mind? The only way I think this could be done is by truncating text in the parser instead of when we are generating the intermediate content string

**Comment by Lomadriel** [comment]
Hum, maybe I miss something, but this can be solved by modifying this function:
https://github.com/jaagr/polybar/blob/d5112c9b663a3c0d8b013f5110bbf448c18c1ffd/src/components/builder.cpp#L547-L562

If I am not wrong, the only tags that are in the label are the formatting ones.
See https://github.com/Lomadriel/polybar/commit/b632f6686db17d896614494503d9cde435adb91b & https://github.com/Lomadriel/polybar/commit/75c712c39ad2381d7417c65f986fa63ac236a985

**Comment by patrick96** [comment]
Well, you'd need to do comprehensive tag parsing inside `get_label_text`. From a quick glance at your code edge cases like `%{A1:echo "}":}` and `ABC%%{F#ffffff}` are not covered.

This would mean we completely parse formatting tags both here and in the parser. We also do parsing of user specified raw tags in the builder currently, but I removed that in https://github.com/jaagr/polybar/pull/1596/commits/2504f0ddabe494297d4138b14907ae597d307adb because it was buggy.

There is a discussion to be had about whether or not it makes sense to parse tokens both before adding them to the output string *and* afterwards in the parser. In some cases it might make sense for example here or when fixing #1701. But if we already parse it beforehand it might make sense to actually store the output string in a more structured way that makes the second parsing unnecessary.

**Comment by Lomadriel** [comment]
> Well, you'd need to do comprehensive tag parsing inside `get_label_text`. From a quick glance at your code edge cases like `%{A1:echo "}":}` and `ABC%%{F#ffffff}` are not covered.

Indeed this is bugged :/

> There is a discussion to be had about whether or not it makes sense to parse tokens both before adding them to the output string _and_ afterwards in the parser. In some cases it might make sense for example here or when fixing #1701. But if we already parse it beforehand it might make sense to actually store the output string in a more structured way that makes the second parsing unnecessary.

It might be a good idea to store the output string in a more structured way and could greatly simplify the manipulation of these strings.
If you want I can try to propose an UML for this big rework and so we will be able to discuss that.

**Comment by fuzxi** [comment]
Could I ask that this be added to the [Known Issues page?](https://github.com/polybar/polybar/wiki/Known-Issues) I just spent a good hour trying to figure out why I was getting a `4 unclosed action blocks` error in a test config that only had two action blocks and this issue didn't come up in my searches until I'd already singled out `label-maxlen`.

**Comment by patrick96** [comment]
[Done](https://github.com/polybar/polybar/wiki/Known-Issues#label-maxlen-truncates-formatting-tags)
